### PR TITLE
Update Resolver to support ARMID refs and namespaces

### PR DESCRIPTION
**Pull request by matthchr**
- Also refactor resolver into genruntime package as in the future it will become a parameter on the ConvertToARM function.

### CODE REVIEW

File: hack/generated/pkg/genruntime/errors.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+Copyright (c) Microsoft Corporation.
```
Comment (matthchr) [inline hack/generated/pkg/genruntime/errors.go:2]: This file was just copied and the error type renamed

File: hack/generated/pkg/genruntime/resource_hierarchy.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+package genruntime
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type ResourceHierarchyRoot string
```
Comment (matthchr) [inline hack/generated/pkg/genruntime/resource_hierarchy.go:15]: This file was just copied -- not sure why diff makes it show all new

File: hack/generated/pkg/genruntime/resource_reference.go [hunk 3]
```diff
@@ -19,17 +22,46 @@ type KnownResourceReference struct {
...
+// TODO: Need to fix up construction of this to include namespace in generated code
 type ResourceReference struct {
-	// The group of the referenced resource.
+	// Group is the Kubernetes group of the resource.
 	Group string `json:"group"`
-	// The kind of the referenced resource.
+	// Kind is the Kubernetes kind of the resource.
 	Kind string `json:"kind"`
-	// The name of the referenced resource.
+	// Namespace is the Kubernetes namesapce of the resource.
```
Comment (matthchr) [inline hack/generated/pkg/genruntime/resource_reference.go:33]: typo: namespace

File: hack/generated/pkg/genruntime/resource_reference.go [hunk 4]
```diff
@@ -19,17 +22,46 @@ type KnownResourceReference struct {
...
+	Namespace string `json:"namespace"`
+	// Name is the Kubernetes name of the resource.
 	Name string `json:"name"`
 
 	// Note: Version is not required here because references are all about linking one Kubernetes
 	// resource to another, and Kubernetes resources are uniquely identified by group, kind, (optionally namespace) and
 	// name - the versions are just giving a different view on the same resource
+
+	// ARMID is a string of the form /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceProviderNamespace}/{resourceType}/{resourceName}. ARMID is mutually exclusive with Group, Kind, Namespace and Name.
```
Comment (matthchr) [inline hack/generated/pkg/genruntime/resource_reference.go:42]: Line too long

File: hack/generated/pkg/genruntime/resource_hierarchy.go (hunk 2 above)
Comment (matthchr) [reply hack/generated/pkg/genruntime/resource_hierarchy.go:15]: Oh, doh -- it didn't exist before and I split it out of the resolver file to make things cleaner. No actual change to the content though.

File: hack/generated/pkg/genruntime/arm_transformer.go [hunk 5]
```diff
@@ -10,9 +10,8 @@ import (
 )
 
 type ToArmConverter interface {
-	// name is the "a/b/c" name for the owner. For example this would be "VNet1" when deploying subnet1
-	// or myaccount when creating Batch pool1
-	ConvertToArm(name string) (interface{}, error)
+	// ConvertToArm converts this to an ARM resource. The name parameter is the fully qualified name of the resource in Azure ("a/b/c").
+	ConvertToArm(name string) (interface{}, error) // TODO: Should be ConvertToARM (same below)
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/genruntime/arm_transformer.go:14]: Does this TODO need to be actioned before you merge?

File: hack/generated/pkg/genruntime/resolver.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	if !ok {
+		// Resource doesn't have a resource ID. This probably means it's not done deploying
+		return "", errors.Errorf("ref %s doesn't have an assigned ARM ID", ref)
+	}
+
+	return id, nil
+}
+
+// ResolveResourceHierarchy gets the resource hierarchy for a given resource. The result is a slice of
+// resources, with the uppermost parent at position 0 and the resource itself at position len(slice)-1
```
Comment (theunrepentantgeek) [inline hack/generated/pkg/genruntime/resolver.go:57]: Nice description of the behaviour. :+1:

File: hack/generated/pkg/genruntime/arm_transformer.go (hunk 5 above)
Comment (matthchr) [reply hack/generated/pkg/genruntime/arm_transformer.go:14]: No, I've got a follow-on PR that does a bunch of `Arm` -> `ARM` renames but I didn't want to mix real changes and renames. Once this is merged I'll send the capitalization fixups for existing methods.

**Review by theunrepentantgeek — APPROVED** [review]

File: hack/generated/pkg/genruntime/arm_transformer.go (hunk 5 above)
Comment (theunrepentantgeek) [reply hack/generated/pkg/genruntime/arm_transformer.go:14]: Good plan.

**Pull request merged**